### PR TITLE
Fix:310 Show messages of that date exclusively

### PR DESCRIPTION
**Pull request by vishwesh3**
PR addressing #310

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@vishwesh3, it looks like you've already signed the Dropbox CLA. Thanks!

**Comment by Sam1301** [comment]
Nice @vishwesh3 , works well :+1:

**Comment by Sam1301** [comment]
just noticed the commit message, it doesn't include the hash `#` before issue number :)

**Comment by kunall17** [comment]
- Include the # in the commit message
- Code not formatted! 

Otherwise LGTM!

**Comment by vishwesh3** [comment]
Done

**Comment by vishwesh3** [comment]
Now messages are filtered from local DB are shown. Filtering will be enhanced once https://github.com/zulip/zulip/issues/3406 is solved. 

https://drive.google.com/open?id=0BwZ3hknwHurYUDJObm0ybDZmUU0

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java [hunk 1]
```diff
@@ -56,23 +47,30 @@ public void writeToParcel(Parcel dest, int flags) {
     @Override
     public Where<Message, Object> modWhere(Where<Message, Object> where)
             throws SQLException {
-        where.like(Message.TIMESTAMP_FIELD, new SelectArg(date));
+        DateUtils.truncate(new Date(), java.util.Calendar.DAY_OF_MONTH);
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: `DateUtils.truncate` returns an rounded date you are not using this why is this here?

https://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/time/DateUtils.html#truncate(java.util.Calendar, int)

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java [hunk 2]
```diff
@@ -56,23 +47,30 @@ public void writeToParcel(Parcel dest, int flags) {
...
+        Calendar calendar = Calendar.getInstance();
+        calendar.setTime(date);
+        calendar.set(Calendar.HOUR_OF_DAY,0);
+        calendar.set(Calendar.MINUTE,0);
+        calendar.set(Calendar.SECOND,0);
+        calendar.set(Calendar.MILLISECOND, 0);
+        where.between(Message.TIMESTAMP_FIELD,calendar.getTime().getTime(),calendar.getTime().getTime()+86400000);
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Declare this 86400000 into a constant, this would help to understand this easier!

Comment (kunall17) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Also format your code!

File: app/src/main/java/com/zulip/android/models/Message.java [hunk 3]
```diff
@@ -104,7 +105,7 @@
     private String subject;
     @SerializedName("timestamp")
     @DatabaseField(columnName = TIMESTAMP_FIELD)
-    private Date timestamp;
+    private long timestamp;
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/models/Message.java]: Changing the column fields isn't the best option, will it still work even for the old message table?

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java (hunk 1 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Removed `DateUtils.truncate`. Forgot to remove it before

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java (hunk 2 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Done

File: app/src/main/java/com/zulip/android/models/Message.java (hunk 3 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/models/Message.java]: As column field is changed so DB version is increased and onUpdate method is called. Thus it working for old messages too. :)

Comment (Sam1301) [inline app/src/main/java/com/zulip/android/models/Message.java]: @vishwesh3 just trying to understand the code, can you please tell why changing the type of timestamp in message model is needed?

Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/models/Message.java]: Let's take a example (assume that back dated messages are stored in local DB (will be fine once https://github.com/zulip/zulip/issues/3406 is done) and today's date is  25/01/2017)

Now when we narrow to date let's say 20/01/2017. 
**Before this PR**
Database query is done with `where.like(Message.TIMESTAMP_FIELD, new SelectArg(date));` (https://github.com/zulip/zulip-android/blob/master/app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java#L59) rather than filtering messages of that specific date it just do nothing. When I debugged I found that list contains messages of all dates, but messages of other days ( except 20/01/2017) are not shown because it returns false on match (https://github.com/zulip/zulip-android/blob/master/app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java#L65). All DB quires are done with limit of 100. 
**Case 1**
So if messages from 20 to 25 are less than 100 thus messages returned by DB query includes messages of 20 and only messages dated 20 are shown with the help of match method. 
**Case 2**
Messages from 20 to 25 are greater than 100 thus messages returned by DB query does not include messages of 20, even if they are stored in the local DB. Also no messages are displayed so scrolling is not possible and no further DB quires takes place.

**Problem** Where clause is not working as per expectation.

**In this PR**
Changed where clause (https://github.com/vishwesh3/zulip-android/blob/narrow/app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java#L57) 
Now DB query returns only those messages whose timestamp is in between narrowed date.

Comment (Sam1301) [inline app/src/main/java/com/zulip/android/models/Message.java]: @vishwesh3 I really appreciate how well you explained your point :) Although I still feel that changing the type of timestamp for a message is unnecessary. Ormlite [where.between](http://ormlite.com/javadoc/ormlite-core/com/j256/ormlite/stmt/Where.html#between(java.lang.String,%20java.lang.Object,%20java.lang.Object)) can take any type of data (in this case Date()) to return a modulated where. So I tried these [changes](https://gist.github.com/Sam1301/90a7271024e2b560d46e7f885a332198) to NarrowFilterByDate.java and it worked for me. The code in my example could obviously be more concise but I guess you get the point :)

Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/models/Message.java]: Thanks @Sam1301 I will play around and will update soon.

Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/models/Message.java]: Done :)

**Comment by vishwesh3** [comment]
Done :) 
Please have a final look.

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java [hunk 4]
```diff
@@ -56,23 +46,37 @@ public void writeToParcel(Parcel dest, int flags) {
...
+        calendar.set(Calendar.MILLISECOND, 0);
+
+        Calendar calendar1 = Calendar.getInstance();
+        calendar1.setTime(date);
+        calendar1.set(Calendar.HOUR_OF_DAY, 0);
+        calendar1.set(Calendar.MINUTE, 0);
+        calendar1.set(Calendar.SECOND, 0);
+        calendar1.set(Calendar.MILLISECOND, 0);
+        calendar1.add(Calendar.DAY_OF_MONTH, 1);
+        where.between(Message.TIMESTAMP_FIELD, calendar.getTime(), calendar1.getTime());
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: there is some code duplication here, can be reduced

Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Done.

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java [hunk 5]
```diff
@@ -56,23 +47,33 @@ public void writeToParcel(Parcel dest, int flags) {
...
+        calendar.setTime(date);
+        calendar.set(Calendar.HOUR_OF_DAY, 0);
+        calendar.set(Calendar.MINUTE, 0);
+        calendar.set(Calendar.SECOND, 0);
+        calendar.set(Calendar.MILLISECOND, 0);
+
+        Calendar calendar1 = Calendar.getInstance();
+        calendar1.setTime(calendar.getTime());
+        calendar1.add(Calendar.DAY_OF_MONTH, 1);
+        where.between(Message.TIMESTAMP_FIELD, calendar.getTime(), calendar1.getTime());
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: ok so here we are creating 2 Calendar instances. I think we could use just one calendar instance
Example:
```
date1 = calendar.getTime();
calendar.add(DAY_OF_MONTH, 1);
date2 = calendar.getTime();
```

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java [hunk 6]
```diff
@@ -3,16 +3,18 @@
...
 
 import org.json.JSONException;
 
 import java.sql.SQLException;
 import java.text.SimpleDateFormat;
+import java.util.Calendar;
 import java.util.Date;
+import java.util.Locale;
 
 public class NarrowFilterByDate implements NarrowFilter {
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Also I just noticed, that we are not using writeToParcel method in NarrowFilterByDate. When we need to pass objects across activities we implement the [Parcelable interface](https://developer.android.com/reference/android/os/Parcelable.html) for those object classes. We are saving NarrowFilter instances in a [bundle](https://github.com/zulip/zulip-android/blob/master/app/src/main/java/com/zulip/android/activities/MessageListFragment.java#L81) to save the state. See [NarrowFilterStream](https://github.com/zulip/zulip-android/blob/master/app/src/main/java/com/zulip/android/filters/NarrowFilterStream.java) for example.

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java (hunk 5 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Done :)

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java [hunk 7]
```diff
@@ -35,44 +37,45 @@ public NarrowFilterByDate(Date date) {
...
-    private static boolean isSameDay(long date1, long date2) {
-        return date1 / 86400000 == date2 / 86400000;
-    }
-
     public int describeContents() {
         return 0;
     }
 
     public void writeToParcel(Parcel dest, int flags) {
+        dest.writeValue(date);
```
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: When you write data to parcel, it is analogous to serialization, so we also need to deserialize the result to get back the NarrowFilterByDate instance in [creator createfromparcel](https://github.com/zulip/zulip-android/blob/master/app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java#L19).

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java [hunk 8]
```diff
@@ -3,23 +3,25 @@
...
+import java.util.Locale;
 
 public class NarrowFilterByDate implements NarrowFilter {
 
     public static final Parcelable.Creator<NarrowFilterByDate> CREATOR = new Parcelable.Creator<NarrowFilterByDate>() {
         public NarrowFilterByDate createFromParcel(Parcel in) {
-
-            return new NarrowFilterByDate();
+            in.setDataPosition(0);
+            return new NarrowFilterByDate((Date) in.readValue(Date.class.getClassLoader()));
```
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Deserialized the result to get back the NarrowFilterByDate instance. :)

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java (hunk 7 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Done :)

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java (hunk 8 above)
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Awesome :) Only one comment though, we don't need to set data position to initial since its already set in the beginning when createFromParcel is called :)

Comment (Sam1301) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: btw just a suggestion, by using writeValue(Date object) we are using general serialization which is kinda slow pref wise against parcelable, and that is the whole point why in android parcelable is preferred over general serialization(https://developer.android.com/reference/android/os/Parcel.html#writeValue(java.lang.Object))

**Comment by Sam1301** [comment]
@vishwesh3 also can you please squash the two commits as they both address the same problem? Maybe even add more description to commit message explaining "why" the change is needed :)

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java (hunk 8 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Used long and created Date with it at the time of deserialization. :)

**Comment by Sam1301** [comment]
@vishwesh3 awesome, can you please fix some typos in the commit message :)

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java (hunk 8 above)
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: cool :)

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java [hunk 9]
```diff
@@ -35,44 +36,45 @@ public NarrowFilterByDate(Date date) {
...
     @Override
     public boolean matches(Message msg) {
-        return isSameDay(date.getTime(), msg.getTimestamp().getTime());
+        return isSameDay(date, new Date(msg.getTimestamp().getTime()));
     }
 
     @Override
     public String getTitle() {
-        return isSameDay(date.getTime(), new Date().getTime()) ? "Today Messages" : "Messages";
+        return isSameDay(date, new Date()) ? "Today Messages" : "Messages";
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: User displayed strings in strings.xml!

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java [hunk 10]
```diff
@@ -95,4 +97,20 @@ public String getJsonFilter() throws JSONException {
...
+     * Checks two dates are of same day or not
+     *
+     * @param date1 Date date1 to be compared with date2
+     * @param date2 Date date2 to be compared with date1
+     * @return boolean
+     */
+    private static boolean isSameDay(Date date1, Date date2) {
+        Calendar calendar = Calendar.getInstance();
+        calendar.setTime(date1);
+        Calendar calendar2 = Calendar.getInstance();
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: This isSameDay will be called very frequently via https://github.com/zulip/zulip-android/blob/a4374c12240d3714710f2dac4376dbecc692a49c/app/src/main/java/com/zulip/android/activities/MessageListFragment.java#L335 
So i'm kinda confused is this getInstance returning a new object? 
If so it would be better to cache these two objects!

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java (hunk 9 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Done :)

File: app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java (hunk 10 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/filters/NarrowFilterByDate.java]: Done :)

**Comment by kunall17** [comment]
@timabbott LGTM!

**Comment by timabbott** [comment]
@vishwesh3 I cleaned up the typos in your commit message and merged this as 41ceacd2522a4b727f7455b74afb4e55d2f401f3.

Thanks for fixing this!!
